File: src/main/scala/com/algorithmlx/inscribers/compact/jei/InscribersJEIPlugin.scala
package com.algorithmlx.inscribers.compact.jei

import com.algorithmlx.inscribers.Constant.reloc
import com.algorithmlx.inscribers.api.JUtils
import com.algorithmlx.inscribers.api.helper.DistHelper
import com.algorithmlx.inscribers.compact.jei.category.InscriberRecipeCategory
import com.algorithmlx.inscribers.init.registry.{InscribersRecipeTypes, Register}
import com.algorithmlx.inscribers.recipe.InscriberRecipe
import mezz.jei.api.registration.{IRecipeCatalystRegistration, IRecipeCategoryRegistration, IRecipeRegistration}
import mezz.jei.api.{IModPlugin, JeiPlugin}
import net.minecraft.client.Minecraft
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation

//noinspection ScalaUnusedSymbol
@JeiPlugin
class InscribersJEIPlugin extends IModPlugin {
  override def getPluginUid: ResourceLocation = reloc("integration_jei")

  override def registerCategories(registration: IRecipeCategoryRegistration): Unit = {
    val guiHelper = registration.getJeiHelpers.getGuiHelper
    registration.addRecipeCategories(new InscriberRecipeCategory(guiHelper))
  }

  override def registerRecipeCatalysts(registration: IRecipeCatalystRegistration): Unit = {
    registration.addRecipeCatalyst(new ItemStack(Register.INSCRIBER_BLOCK.get()), reloc("inscriber"))
  }

  override def registerRecipes(registration: IRecipeRegistration): Unit = {
    if (DistHelper.isLogicalClient) {
      val level = Minecraft.getInstance().level
      if (level == null) return
      val recipeManager = level.getRecipeManager
      val recipes = JUtils.getRecipe(recipeManager)

      registration.addRecipes(recipes.values(), reloc("inscriber"))
    }
  }
}

File: src/main/java/com/algorithmlx/inscribers/compact/jei/InscribersJEIPlugin.kt
package com.algorithmlx.inscribers.compact.jei

import com.algorithmlx.inscribers.api.isLogicalClient
import com.algorithmlx.inscribers.api.mc
import com.algorithmlx.inscribers.compact.jei.category.InscriberRecipeCategory
import com.algorithmlx.inscribers.init.registry.InscribersRecipeTypes
import com.algorithmlx.inscribers.init.registry.Register
import com.algorithmlx.inscribers.reloc
import mezz.jei.api.IModPlugin
import mezz.jei.api.JeiPlugin
import mezz.jei.api.registration.IRecipeCatalystRegistration
import mezz.jei.api.registration.IRecipeCategoryRegistration
import mezz.jei.api.registration.IRecipeRegistration
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation

@JeiPlugin
class InscribersJEIPlugin: IModPlugin {
    override fun getPluginUid(): ResourceLocation = reloc("jei")

    override fun registerCategories(registration: IRecipeCategoryRegistration) {
        val guiHelper = registration.jeiHelpers.guiHelper
        registration.addRecipeCategories(InscriberRecipeCategory(guiHelper))
    }

    override fun registerRecipeCatalysts(registration: IRecipeCatalystRegistration) {
        registration.addRecipeCatalyst(ItemStack(Register.basicInscriberBlock.get()), reloc("inscriber"))
        registration.addRecipeCatalyst(ItemStack(Register.advancedInscriber.get()), reloc("inscriber"))
        registration.addRecipeCatalyst(ItemStack(Register.improvedInscriber.get()), reloc("inscriber"))
        registration.addRecipeCatalyst(ItemStack(Register.eliteInscriber.get()), reloc("inscriber"))
        registration.addRecipeCatalyst(ItemStack(Register.perfectInscriber.get()), reloc("inscriber"))
        registration.addRecipeCatalyst(ItemStack(Register.maximizedInscriber.get()), reloc("inscriber"))
    }

    override fun registerRecipes(registration: IRecipeRegistration) {
        if (isLogicalClient()) {
            val level = mc.level ?: return
            val recipeManager = level.recipeManager
            val recipe = recipeManager.byType(InscribersRecipeTypes.matrixInscriberRecipe)

            registration.addRecipes(recipe.values, reloc("inscriber"))
        }
    }
}